include/seastar/core/internal/io_request.hh

```diff
@@ -266,6 +266,18 @@ public:
         return io_request(operation_tag<operation::cancel>{}, fd, addr);
     }
 
+    io_request(io_request&& other) noexcept
+        : _op(other._op) {
+        std::memcpy(static_cast<void*>(this), static_cast<void*>(&other), sizeof(*this));
```

Comment:
I'm sure violates some language rule. It also pessimizes the circular_buffer use case.

Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.

Reply:
>  It also pessimizes the circular_buffer use case.

 i am not sure what exactly you are referring. is it 

- `io_request(io_request&& other)` operator?  the default-generated one is deleted because otherwise it'd be ill-formed. see https://app.circleci.com/pipelines/github/tchaikov/seastar/170/workflows/c025345c-cf68-4005-a92b-cb5cb5e6a809/jobs/1994 . 
- the `std::memcpy()` call? i think all members variable of the union members are `std::is_trivially_copyable<>`, so we should be able to `memcpy()` them, right?
-  `static_cast<void*>()`? without it, we'd have `-Wclass-memaccess`. like:
   ```
   ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
   ‘ io_request’ with no trivial copy-assignment;
   use copy-initialization instead [-Wclass-memaccess]
   ```

> Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.

i think what @xemul is after is `const` specifier on the members instead. Pavel, am i right? and i am afraid that the value type of `circular_buffer<Value>` can be a const type, if that's one case of the "everything".

Reply:
> > It also pessimizes the circular_buffer use case.
> 
> i am not sure what exactly you are referring. is it
> 
> * `io_request(io_request&& other)` operator?  the default-generated one is deleted because otherwise it'd be ill-formed. see https://app.circleci.com/pipelines/github/tchaikov/seastar/170/workflows/c025345c-cf68-4005-a92b-cb5cb5e6a809/jobs/1994 .


User-defined move constructors pessimizes circular_buffer since it now needs to call it.

Without the user-defined constructor, it can just memcpy stuff itself.

> * the `std::memcpy()` call? i think all members variable of the union members are `std::is_trivially_copyable<>`, so we should be able to `memcpy()` them, right?

But memcpy to const members is illegal. The reinterpret_cast is also secretly a const_cast.

> * `static_cast<void*>()`? without it, we'd have warnings from `-Wclass-memaccess`. like:
>   ```
>   ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
>   ‘ io_request’ with no trivial copy-assignment;
>   use copy-initialization instead [-Wclass-memaccess]
>   ```
> 
> > Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.
> 
> i think what @xemul is after is `const` specifier on the members instead. Pavel, am i right? and i am afraid that the value type of `circular_buffer<Value>` can be a const type, if that's one case of the "everything".

Well, it's a struct. I don't see why we have to qualify members of a struct as const, it's not something typically done in C++ except when there's some special member that must be const.

Reply:
> > > It also pessimizes the circular_buffer use case.
> > 
> > 
> > i am not sure what exactly you are referring. is it
> > 
> > * `io_request(io_request&& other)` operator?  the default-generated one is deleted because otherwise it'd be ill-formed. see https://app.circleci.com/pipelines/github/tchaikov/seastar/170/workflows/c025345c-cf68-4005-a92b-cb5cb5e6a809/jobs/1994 .
> 
> User-defined move constructors pessimizes circular_buffer since it now needs to call it.
> 
> Without the user-defined constructor, it can just memcpy stuff itself.

agreed. this is sad. without user-defined constructor. it'd be difficult to add the `const` specifiers.

> 
> > * the `std::memcpy()` call? i think all members variable of the union members are `std::is_trivially_copyable<>`, so we should be able to `memcpy()` them, right?
> 
> But memcpy to const members is illegal. The reinterpret_cast is also secretly a const_cast.

yes. the assignment operator just overwrite the immutable members. this practically defeats the purpose of `const`.

> 
> > * `static_cast<void*>()`? without it, we'd have warnings from `-Wclass-memaccess`. like:
> >   ```
> >   ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
> >   ‘ io_request’ with no trivial copy-assignment;
> >   use copy-initialization instead [-Wclass-memaccess]
> >   ```
> > 
> > > Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.
> > 
> > 
> > i think what @xemul is after is `const` specifier on the members instead. Pavel, am i right? and i am afraid that the value type of `circular_buffer<Value>` can be a const type, if that's one case of the "everything".
> 
> Well, it's a struct. I don't see why we have to qualify members of a struct as const, it's not something typically done in C++ except when there's some special member that must be const.

ahh, i see your point. it's not specifically a certain (set) of member variables in the op. it's about the whole struct. as the op as a whole is immutable. makes sense to me! but i don't have a quick answer which allows us to make this explicit by adding a `const` somewhere.

maybe we can get the first one reviewed and merged?

i am going to drop the 2nd commit and the 3rd one once @xemul acks this.

Reply:
instead of dropping the 2nd commit and the 3rd commit, i followed @avikivity 's comment:
 
> Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.

and replaced them with a better solution -- to literally making everything `const`.